File: tracker/src/Type3830.f90
Subroutine Type3830
  !------------------------------------------------------------------------------------
  !    DESCRIPTION
  !------------------------------------------------------------------------------------
  ! Subroutine Type3830 writes TRNSYS simulation timing information to a
  ! `***.tmp` file for external monitoring and post-processing.
  !
  ! The subroutine records:
  !   - Current simulation time (TIME)
  !   - Simulation start time (START)
  !   - Simulation stop time (STOP)
  !   - Current timestep size (STEP)
  !
  ! Version History:
  !   2026-01-01 – A. Lachance: Original implementation.
  !   2026-06-20 – A. Lachance: Modified description
  !
  ! This subroutine can be exported for use in external DLLs.
  !DEC$ATTRIBUTES DLLEXPORT :: TYPE3830
  !------------------------------------------------------------------------------------
  !    VARIABLES
  !------------------------------------------------------------------------------------
  Use TrnsysConstants
  Use TrnsysFunctions
  Implicit None

  Double Precision :: timestep, time
  Double Precision :: sim_start, sim_stop
  Integer :: currentUnit, currentType
  Integer :: lu_print, ios
  Logical :: lu_ok
  Double Precision :: print_interval, next_print_time

  Character(len=maxLabelLength) :: fileName
  Character(len=32) :: LUStr
  Character(len=256) :: Message

  !------------------------------------------------------------------------------------
  !    INITIALIZATION
  !------------------------------------------------------------------------------------
  ! Get global TRNSYS simulation variables
  time = getSimulationTime()
  timestep = getSimulationTimeStep()
  sim_start = getSimulationStartTime()
  sim_stop = getSimulationStopTime()
  currentUnit = getCurrentUnit()
  currentType = getCurrentType()

  !------------------------------------------------------------------------------------
  !    VERSION CHECK
  !------------------------------------------------------------------------------------
  If (getIsVersionSigningTime()) Then
    Call SetTypeVersion(17)
    Return
  End If

  !------------------------------------------------------------------------------------
  !    FINAL CALL HANDLING
  !------------------------------------------------------------------------------------
  If (getIsLastCallofSimulation()) Then
    fileName = getLUFileName(lu_print)
    REWIND (lu_print)
    Write (lu_print, '(F0.10,", ",F0.10,", ",F0.10,", ",F0.10)', ERR=200) time, sim_start, sim_stop, timestep
    Close (lu_print)
    Return
  End If

  !------------------------------------------------------------------------------------
  !    FINAL STEP HANDLING
  !------------------------------------------------------------------------------------
  If (getIsEndOfTimestep()) Then
    next_print_time = getStaticArrayValue(1)

    If (time + timestep/2.d0 >= next_print_time) Then
      fileName = getLUFileName(lu_print)

      REWIND (lu_print)
      Write (lu_print, '(F0.10,", ",F0.10,", ",F0.10,", ",F0.10)', ERR=200) time, sim_start, sim_stop, timestep

      next_print_time = next_print_time + print_interval
      Call SetStaticArrayValue(1, next_print_time)
    End If

    Return
  End If

  !------------------------------------------------------------------------------------
  !    TYPE INITIALIZATION
  !------------------------------------------------------------------------------------
  If (getIsFirstCallofSimulation()) Then
    ! TRNSYS Engine Type Calls
    Call SetNumberofParameters(2)
    Call SetNumberofInputs(0)
    Call SetNumberofDerivatives(0)
    Call SetNumberofOutputs(0)
    Call SetIterationMode(5)
    Call SetNumberStoredVariables(1, 0)
    Call SetNumberofDiscreteControls(0)

    ! TRNSYS Input and Output Units
    ! Nothing

    Return
  End If

  !------------------------------------------------------------------------------------
  !    INITIAL VALUE SETTING
  !------------------------------------------------------------------------------------
  If (getIsStartTime()) Then

    print_interval = getParameterValue(2)
    If (print_interval <= 0) Call FoundBadParameter(2, "Fatal", "Printing interval must be larger than 0")
      If (print_interval < timestep) Call FoundBadParameter(2,"Fatal","Printing interval must be larger or equal than Simulation time step")
    Call SetStaticArrayValue(1, print_interval)

    lu_print = JFIX(getParameterValue(1) + 0.01)
    lu_ok = .false.

    ! If a positive logical unit is specified
    If (lu_print > 0) Then
      ! Reject units less than 10
      If (lu_print >= 10) Then
        lu_ok = LogicalUnitIsOpen(lu_print)
      End If

      ! If no valid logical unit specified, get default listing unit
    Else If (lu_print < 0) Then
      lu_print = getListingFileLogicalUnit()
      lu_ok = .true.
    End If

    ! If logical unit is still not valid, report fatal error
    If (.not. lu_ok) Then
      Write (LUStr, *) lu_print
      Message = "Logical Unit Number = "//TRIM(ADJUSTL(LUStr))
      Call Messages(431, Message, "FATAL", CurrentUnit, CurrentType)
      Return
    End If

    fileName = getLUFileName(lu_print)
    Open (Unit=lu_print, File=fileName, Status="Replace", Action="Write", IOSTAT=ios)

    Return
  End If

  !------------------------------------------------------------------------------------
  !    RE-READ PARAMETERS
  !------------------------------------------------------------------------------------
  If (getIsReReadParameters()) Then
    lu_print = JFIX(getParameterValue(1) + 0.01)
    print_interval = getParameterValue(2)
    fileName = getLUFileName(lu_print)
  End If

  !------------------------------------------------------------------------------------
  !    MAIN TYPE CODE
  !------------------------------------------------------------------------------------
200 Call Messages(431, Message, "FATAL", CurrentUnit, CurrentType)

END SUBROUTINE Type3830

File: type3830/src/type3830.nim
# =============================================================================
# DESCRIPTION
# =============================================================================
# Subroutine Type3830 writes TRNSYS simulation timing information to a
# `***.tmp` file for external monitoring and post-processing.
#
# The subroutine records:
#   - Current simulation time (TIME)
#   - Simulation start time (START)
#   - Simulation stop time (STOP)
#   - Current timestep size (STEP)
#
# Only one instance of Type3830 may be used in a simulation.
#
# Version history:
#   2026-01-01 – A. Lachance: Original implementation.
#   2026-06-20 – A. Lachance: Updated description.
#   2026-08-13 – A. Lachance: Converted to Nim.
#   2026-08-15 – A. Lachance: Optimized runtime access and file writing.

import std/strformat

import bindings/trnsys

var
  timingFile: File
  printInterval: cdouble
  nextPrintTime: cdouble
  simulationStart: cdouble
  simulationStop: cdouble
  simulationStep: cdouble

proc discardTimingFile() =
  if timingFile != nil:
    try:
      timingFile.close()
    except CatchableError:
      discard
    timingFile = nil

proc writeTimingRecord(time: cdouble; closeAfter: bool = false): bool =
  result = false

  if timingFile == nil:
    return false

  try:
    timingFile.setFilePos(0)
    timingFile.write(&"{time:22.10f}, {simulationStart:22.10f}, {simulationStop:22.10f}, {simulationStep:22.10f}\n")

    if closeAfter:
      timingFile.close()
      timingFile = nil
    else:
      timingFile.flushFile()

    result = true
  except CatchableError:
    let errorMessage = getCurrentExceptionMsg()

    discardTimingFile()
    trnsysFatal(431, "Type3830 file operation failed: " & errorMessage)

proc TYPE3830() {.cdecl, exportc: "TYPE3830", dynlib.} =
  # ------------------------------------------------------------
  # VERSION
  # ------------------------------------------------------------
  if getIsVersionSigningTime() != 0:
    setTypeVersion(17)
    return

  # ------------------------------------------------------------
  # END OF SIMULATION
  # ------------------------------------------------------------
  if getIsLastCallOfSimulation() != 0:
    discard writeTimingRecord(getSimulationTime(), true)
    return

  # ------------------------------------------------------------
  # END OF TIMESTEP
  # ------------------------------------------------------------
  if getIsEndOfTimestep() != 0:
    if timingFile == nil:
      return

    let time = getSimulationTime()

    if time + simulationStep * 0.5 < nextPrintTime:
      return

    if writeTimingRecord(time):
      nextPrintTime += printInterval

    return
  # ------------------------------------------------------------
  # INITIALIZATION
  # ------------------------------------------------------------
  if getIsFirstCallOfSimulation() != 0:
    setNumberOfParameters(2)
    setNumberOfInputs(0)
    setNumberOfDerivatives(0)
    setNumberOfOutputs(0)
    setIterationMode(5)
    setNumberStoredVariables(0, 0)
    setNumberOfDiscreteControls(0)
    return

  # ------------------------------------------------------------
  # START OF SIMULATION
  # ------------------------------------------------------------
  if getIsStartTime() != 0:
    simulationStart = getSimulationStartTime()
    simulationStop = getSimulationStopTime()
    simulationStep = getSimulationTimeStep()
    printInterval = getParameterValue(2)

    if printInterval <= 0.0:
      foundBadParameter(2, "Fatal", "Printing interval must be greater than 0")
      return

    if printInterval < simulationStep:
      foundBadParameter(2, "Fatal", "Printing interval must be greater than or equal to the simulation timestep")
      return

    nextPrintTime = simulationStart + printInterval

    let luPrint = cint(getParameterValue(1) + 0.01)

    if luPrint < 10:
      trnsysFatal(431, "Logical unit must be 10 or greater; got " & $luPrint)
      return

    if not logicalUnitIsOpen(luPrint):
      trnsysFatal(431, "Logical unit " & $luPrint & " is not open")
      return

    let fileName = getLUFileName(luPrint)

    if fileName.len == 0:
      trnsysFatal(431, "No file is assigned to logical unit " & $luPrint)
      return

    try:
      discardTimingFile()
      closeFileIVF(luPrint)

      if not open(timingFile, fileName, fmWrite):
        timingFile = nil
        trnsysFatal(431, "Unable to open timing file: " & fileName)
        return

      discard writeTimingRecord(getSimulationTime())
    except CatchableError:
      let errorMessage = getCurrentExceptionMsg()

      discardTimingFile()
      trnsysFatal(431, "Type3830 file operation failed: " & errorMessage)

    return

  # ------------------------------------------------------------
  # RE-READ PARAMETERS
  # ------------------------------------------------------------
  if getIsReReadParameters() != 0:
    discard

  # ------------------------------------------------------------
  # MAIN CODE
  # ------------------------------------------------------------
  # This type has no main calculation.
